File: src/yx/primitives/data_chunking.py
"""
Data chunking utilities.

Traceability:
- protocol/specs/architecture/protocol-layers.md (Chunking)
- protocol/specs/technical/yx-protocol-spec.md (Chunking Algorithm)
- protocol/specs/technical/default-values.md (chunk_size = 1024)
"""

from typing import List


def chunk_data(data: bytes, chunk_size: int = 1024) -> List[bytes]:
    """
    Split data into fixed-size chunks.

    Args:
        data: Data to split
        chunk_size: Maximum size per chunk (default: 1024)

    Returns:
        List of byte chunks (at least one, even for empty data)

    Traceability:
    - protocol/specs/technical/default-values.md (chunk_size = 1024)
    """
    if chunk_size <= 0:
        raise ValueError("chunk_size must be positive")

    if not data:
        return [b""]

    chunks = []
    for i in range(0, len(data), chunk_size):
        chunks.append(data[i:i + chunk_size])
    return chunks


def unchunk_data(chunks: List[bytes]) -> bytes:
    """
    Reassemble chunks into original data.

    Args:
        chunks: Ordered list of chunks

    Returns:
        Reassembled bytes

    Traceability:
    - protocol/specs/architecture/protocol-layers.md (Reassembly Algorithm)
    """
    return b"".join(chunks)

File: swift/builds/swift-impl/Sources/YXProtocol/Primitives/DataChunking.swift
import Foundation

// Implements: protocol/specs/technical/yx-protocol-spec.md § Data Chunking

// MARK: - Data Chunking

extension Data {
    /// Split data into fixed-size chunks.
    /// - Parameter chunkSize: Max bytes per chunk (default 1024)
    /// - Returns: Array of chunks
    public func chunked(size chunkSize: Int = 1024) -> [Data] {
        guard !isEmpty else { return [] }
        guard chunkSize > 0 else { return [self] }
        var chunks: [Data] = []
        var offset = 0
        while offset < count {
            let end = Swift.min(offset + chunkSize, count)
            chunks.append(self[offset..<end])
            offset = end
        }
        return chunks
    }
}

// MARK: - BufferEntry

public enum ChunkingError: Error, CustomStringConvertible {
    case incompleteMessage
    case missingChunk(UInt32)

    public var description: String {
        switch self {
        case .incompleteMessage: return "Not all chunks received"
        case .missingChunk(let i): return "Missing chunk \(i)"
        }
    }
}

/// Buffer for incomplete chunked messages.
public struct BufferEntry {
    public var chunks: [UInt32: Data]
    public let totalChunks: UInt32
    public let timestamp: Date

    public init(chunks: [UInt32: Data], totalChunks: UInt32, timestamp: Date) {
        self.chunks = chunks
        self.totalChunks = totalChunks
        self.timestamp = timestamp
    }

    public var isComplete: Bool { chunks.count == Int(totalChunks) }

    /// Reassemble chunks in order.
    public func reassemble() throws -> Data {
        guard isComplete else { throw ChunkingError.incompleteMessage }
        var result = Data()
        for i in 0..<totalChunks {
            guard let chunk = chunks[i] else { throw ChunkingError.missingChunk(i) }
            result.append(chunk)
        }
        return result
    }
}
